Converts paragraphs to headings with keyboard shortcuts

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Closes #11173


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
For feature parity with the classic editor in WordPress but also because they're a great set of shortcuts to have.


## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Not sure I am doing it right, because:

- This PR adds a new dependency to the block library package for this implementation. 
- Should the block handle this sort of shortcut or should the editor, or the editor implementation (post/site)?
- Is this repetitive way enough? Should we update how transforms work (in particular make `switchToBlockType` receive some 
attributes ...)?

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Create a post
2. Write a paragraph
3. Press Ctrl + Alt (Option) + [1-6]
4. Observe magic transform to heading that is the level of [1-6]

## Screenshots



https://user-images.githubusercontent.com/107534/194541040-052f0bd2-f586-4908-afa9-ebeb14407257.mp4


## To do

- [x] Fix some typos in the code
- [x] Make sure only the selected block is transformed
- [x] Add a combo+0 to go back to paragraph
- [x] Register the shortcuts for the heading block as well to switch between levels
The mechanism in the heading block would be different as it's only a "level" change. So maybe a separate PR?
- [x] Refactor the registration/handler into a hook
- [x] Make sure the shortcuts appear in the Keyboard shortcuts help modal
- [x] Figure out what to use for Windows browsers as `access+[1-6]` is for pinned items
- [x] Refactor to remove the changes to RichText

draganescu committed:
adds shortcuts for heading transforms to the paragraph block

getdave commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -52,6 +58,115 @@ function hasDropCapDisabled( align ) {
 	return align === ( isRTL() ? 'left' : 'right' ) || align === 'center';
 }
 
+function ParagrapgShortcuts( { attributes, clientId } ) {
```
```suggestion
function ParagraphShortcuts( { attributes, clientId } ) {
```

talldan replied:
It'd be good to move this component into a separate file. It could also just be a hook.

getdave commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -86,6 +201,10 @@ function ParagraphBlock( {
 
 	return (
 		<>
+			<ParagrapgShortcuts
```
```suggestion
			<ParagraphShortcuts
```

getdave reviewed (commented):
If these shortcuts are specific to the block then it seems reasonable to register them on the block itself.

I wonder:

- will this register the shortcuts repeatedly for each instance of the block?
- do you need to have a method of _de_registering the shortcuts if the block is not in use (see [this example](https://github.com/WordPress/gutenberg/blob/260bcb583652ee60c181073dec62f8163a386b74/packages/block-editor/src/components/keyboard-shortcuts/index.js#L13))?

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -52,6 +58,115 @@ function hasDropCapDisabled( align ) {
+		replaceBlocks(
+			clientId,
+			createBlock( 'core/heading', { level: 1, content, align } )
+		);
```
At the moment if you have multiple paragraphs in a post, this will transform every paragraph, and it should only work for the selected block. 😄

draganescu replied:
Hahah, great.

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -52,6 +58,115 @@ function hasDropCapDisabled( align ) {
+	registerShortcut( {
+		name: 'core/blocks/paragraph/transform-to-heading-1',
+		category: 'block-library',
+		description: __( 'Transform to heading.' ),
```
```suggestion
		description: __( 'Transform a paragraph to heading level 1.' ),
```
I think this should be a user facing description. Ideally it shows in the Keyboard shortcuts modal, but that's not happening for some reason. Possibly because the `category` isn't one that exists.

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -52,6 +58,115 @@ function hasDropCapDisabled( align ) {
+		name: 'core/blocks/paragraph/transform-to-heading-1',
+		category: 'block-library',
+		description: __( 'Transform to heading.' ),
+		keyCombination: {
+			modifier: 'access',
+			character: '1',
+		},
```
On Windows, Ctrl+Alt+{number} keys are used to open the pinned apps in the corresponding position one the taskbar, so I don't think this is a good shortcut combination.

draganescu replied:
That is what @ellatrix suggested. On OSX this works swell. @t-hamano you've done some windows compat fixes, do you have any suggestion as to what could work without conflicts on Wirdows?

t-hamano replied:
In Windows, I think the shortcut to open a pinned app is "Windows logo key + number" (Please see "Windows logo key keyboard shortcuts" section).
https://support.microsoft.com/en-us/windows/keyboard-shortcuts-in-windows-dcc61a57-8ff0-cffe-9796-cb9706c75eec

Also, this is very complicated, but when the modifier is `access`, it means `CTRL + ALT` on Mac and `SHIFT + ALT` on Windows.
https://github.com/WordPress/gutenberg/blob/b66257558a7c7d3bffd88c5c9d3e4e75a0c7e828/packages/keycodes/src/index.js#L164

I report that the current implementation works as expected on Windows 🙂

<details>
<summary>Screencast on Windows</summary>

https://user-images.githubusercontent.com/54422211/194326122-d5571f80-7867-487a-9c2a-c816f2adecde.mp4

</details>

talldan replied:
> In Windows, I think the shortcut to open a pinned app is "Windows logo key + number" (Please see "Windows logo key keyboard shortcuts" section).

You're right, I forgot that my keyboard has alt and windows keys swapped over because I usually work on a mac 😆

talldan reviewed (commented):
It's a nice convenience feature. 👍 

The thing that stood out in testing is there's no easy way to reverse the action (other than undo). Maybe a modifier+`0` could be an option for transforming back to a paragraph.

Also, once you've converted to a heading it should be possible to continue using the shortcut to switch between other heading levels, so I think the same shortcuts could be added to the heading block.

Adding this code directly to the blocks is a reasonable first iteration. I think there might be other options that are worth exploring. 


<details>
<summary>
The shortcuts are really just block transforms, so perhaps it should be possible to specify a keyboard shortcut for a transform and have the shortcut display in the transform menu:
</summary>

![Screen Shot 2022-10-05 at 11 57 10 am](https://user-images.githubusercontent.com/677833/193978342-148f056a-bf17-430a-8bb5-f34814448030.png)

</details>

I feel like that would be an interesting solution, but not without problems - at the moment the only transform between a paragraph and heading is for the default Heading block which is always a `<h2>` level. There's some overlap here with #32472 and #33811, which suggests showing each heading level in the quick inserter and there was a previous attempt to add the levels as variations, but seems it was abandoned. There's lots of opportunities here, but it's just a matter of finding the right solution.

draganescu commented:
> * will this register the shortcuts repeatedly for each instance of the block?

Yes. This is one reason why letting the block handle this doesn't look like the best option possible. It's possible to avoid registering every time sure! But the naive implementation shows some potential bad architecture.

Also, I don't understand the example link in relation to deregistering. As an aside it's unlikely the paragraph block could be unused.

getdave commented:
> Also, I don't understand the example link in relation to deregistering. As an aside it's unlikely the paragraph block could be unused.

There's a `useEffect` which returns a callback "cleanup" function which de-registers the shortcut.

Could we register the shortcut at an editor level conditionally based on whether paragraph and heading blocks are available?

t-hamano commented:
My idea is to implement it in the `block-editor` package itself.
It would solve the problem of multiple shortcut registrations and additional dependencies.

Register shortcuts here:
https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/keyboard-shortcuts/index.js

Like this:

```javascript
[ 1, 2, 3, 4, 5, 6 ].forEach( ( level ) => {
	registerShortcut( {
		name: `core/block-editor/transform-paragraph-to-heading-${ level }`,
		category: 'block',
		keyCombination: {
			modifier: 'access',
			character: level.toString(),
		},
	} );
} );
```

And catch the shortcuts here:
https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-tools/index.js

Like this:

```javascript
[ 1, 2, 3, 4, 5, 6 ].forEach( ( level ) => {
	if (
		isMatch(
			`core/block-editor/transform-paragraph-to-heading-${ level }`,
			event
		)
	) {
		console.log( `transform paragraph to heading H${ level }` );
		// Do transform...
	}
} );
```

draganescu committed:
Implements feedback on hook location, UX and usage.

- Moves the functionality from the paragraph block to the block editor package
- Implements the shortcuts for paragraph and heading allowing cycling between them
- Refactors the code in a less repetitive shape
- Adds access+0 as a way to convert headings to paragraphs
- Implements cycling through heading levels via access+[1-6]
- Registers shortcuts only once
- Adds a new prop to RichText allowing for onKeyDown custom handlers
- Removes the keyboard-shortcuts dependency from block libary

Co-authored-by: Daniel Richards <677833+talldan@users.noreply.github.com>
Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>
Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

draganescu commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -172,6 +174,7 @@ function ParagraphBlock( {
 								'Empty block; start writing or type forward slash to choose a block'
 						  )
 				}
+				onKeyDown={ semanticShorcutsHandler }
```
I think this messes up other rich text shortcuts.

draganescu commented:
Converted this to draft as it is currently breaking other rich text keyboard interactions.

draganescu commented:
I think this should be refactored in a way in which it is the post editor's job to handle these shortcuts. I'll refactor to try and replicate in the post editor how the block editor handles block level shortcuts.

draganescu committed:
Refactors so that setting up the text level shortcuts is handled by the post editor.

draganescu committed:
removes the old hook

draganescu committed:
adds the text level keyboard shortcuts to the help modal of the post editor

talldan commented on packages/edit-post/src/hooks/useTextLevelShortcuts.js:
```diff
@@ -0,0 +1,73 @@
+
+	return [
+		() => {
+			[ 1, 2, 3, 4, 5, 6 ].forEach( ( level ) => {
+				registerShortcut( {
+					name: `core/block-editor/transform-paragraph-to-heading-${ level }`,
+					category: 'block-library',
+					description: __( 'Transform paragraph to heading.' ),
+					keyCombination: {
+						modifier: 'access',
+						character: `${ level }`,
+					},
+				} );
+			} );
+			registerShortcut( {
+				name: `core/block-editor/transform-heading-to-paragraph`,
+				category: 'block-library',
+				description: __( 'Transform heading to paragraph.' ),
+				keyCombination: {
+					modifier: 'access',
+					character: `0`,
+				},
+			} );
```
I think it'd be ok to inline this next to where all the other `registerShortcut` calls are. It's not a lot of code and very easy to understand.

draganescu replied:
I thought that file is long enough :D

talldan commented on packages/edit-post/src/hooks/useTextLevelShortcuts.js:
```diff
@@ -0,0 +1,73 @@
+				getSelectedBlockClientId,
+				getBlockAttributes,
+			} = select( blockEditorStore );
+			const currentClientId = getSelectedBlockClientId();
```
This returns `null` in case of a multiselection, so I think it'd be good to return early when `currentClientId === null`.

talldan commented on packages/edit-post/src/hooks/useTextLevelShortcuts.js:
```diff
@@ -0,0 +1,73 @@
+/**
```
The file name should be kebab case instead of camel case 😄 

(if you decide to keep it)

talldan commented on packages/edit-post/src/components/keyboard-shortcuts/index.js:
```diff
@@ -200,6 +205,35 @@ function KeyboardShortcuts() {
 		setIsListViewOpened( ! isListViewOpened() )
 	);
 
+	useShortcut(
+		'core/block-editor/transform-paragraph-to-heading-1',
+		( event ) => handleTextLevelShortcut( event, 1 )
+	);
+	useShortcut(
+		'core/block-editor/transform-paragraph-to-heading-2',
+		( event ) => handleTextLevelShortcut( event, 2 )
+	);
+	useShortcut(
+		'core/block-editor/transform-paragraph-to-heading-3',
+		( event ) => handleTextLevelShortcut( event, 3 )
+	);
+	useShortcut(
+		'core/block-editor/transform-paragraph-to-heading-4',
+		( event ) => handleTextLevelShortcut( event, 4 )
+	);
+	useShortcut(
+		'core/block-editor/transform-paragraph-to-heading-5',
+		( event ) => handleTextLevelShortcut( event, 5 )
+	);
+	useShortcut(
+		'core/block-editor/transform-paragraph-to-heading-6',
+		( event ) => handleTextLevelShortcut( event, 6 )
+	);
```
I think this could technically put this into a loop as well without breaking the rule of hooks (because the loop is based on a constant, it will always execute the same way every time).

You could then avoid having to export `handleTextLevelShortcut` from a hook, and have it work the same way as the other `useShortcut` calls where all the logic is inline.

draganescu replied:
I felt I needed permission to ~~break~~ bend the holy react rules

talldan commented on packages/edit-post/src/hooks/useTextLevelShortcuts.js:
```diff
@@ -0,0 +1,73 @@
+				getBlockName,
+				getSelectedBlockClientId,
+				getBlockAttributes,
+			} = select( blockEditorStore );
```
Interesting that you use `select` here instead of `useSelect`.

talldan commented on packages/edit-post/src/hooks/useTextLevelShortcuts.js:
```diff
@@ -0,0 +1,73 @@
+			} = select( blockEditorStore );
+			const currentClientId = getSelectedBlockClientId();
+			if (
+				getBlockName( currentClientId ) !== 'core/paragraph' &&
+				getBlockName( currentClientId ) !== 'core/heading'
```
You could assign `getBlockName( currentClientId )` to a variable here to avoid the repeated call.

talldan commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js:
```diff
@@ -36,4 +36,14 @@ export const textFormattingShortcuts = [
 		description: __( 'Make the selected text inline code.' ),
 	},
+	{
+		keyCombination: { modifier: 'access', character: 'Level' },
```
I wonder if instead of `'Level'` it could be `'1-6'`

draganescu replied:
For some reason it can't

draganescu replied:
Oh the thing again, in the function that displays shortcut, I hit that strip regexp back when added `[[` :)

talldan commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js:
```diff
@@ -36,4 +36,14 @@ export const textFormattingShortcuts = [
+	{
+		keyCombination: { modifier: 'access', character: 'Level' },
+		description: __(
+			'Convert the current paragraph to a heading of level 1-6.'
```
It also converts headings so it could say 'paragraph or heading'

talldan commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js:
```diff
@@ -36,4 +36,14 @@ export const textFormattingShortcuts = [
+		),
+	},
+	{
+		keyCombination: { modifier: 'access', character: '0' },
```
Nitpick, I think this could be put above the other shortcut, so that it's in numerical order (0 and then 1-6)

talldan reviewed (commented):
Works really well now. I think a few aspects of the code could be tidied up, but it's not far off being ready. Nice work!

draganescu committed:
Removes the hook based approach and inlines every shortcut definition and handler into edit post shortcut conf file.

Co-authored-by: Daniel Richards <677833+talldan@users.noreply.github.com>

draganescu committed:
updated snapshot for keyboard help modal

draganescu committed:
Refactores the shortcut handler to be a bit quicker returning early for multiselection and avoiding an extra state select.

Co-authored-by: Daniel Richards <677833+talldan@users.noreply.github.com>

draganescu commented:
Cool review @talldan 👏🏻  Applied all feedback, ready for a new check.

bph commented:
So I tested this as fan of keyboard shortcut so yah for @draganescu to tackle. 
I use to use the ### method on paragraphs that I wanted to be come H3 so at first I was kind of, do we really need this. 
After testing it, we definitely need this. I like that is quite simple to use, it supposed to work on paragraphs, but in actually also works on heading blocks, too. 

Nice, it already listed in the Keyboard shortcuts page! Love it. :-)

scruffian commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/test/__snapshots__/index.js.snap:
```diff
@@ -821,6 +821,76 @@ exports[`KeyboardShortcutHelpModal should match snapshot when the modal is activ
+            class="edit-post-keyboard-shortcut-help-modal__shortcut-term"
+          >
+            <kbd
+              aria-label="Shift + Alt + 0"
```
Isn't it Ctrl + Alt + 0?

draganescu replied:
Hmm maybe that is not access?

scruffian commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/test/__snapshots__/index.js.snap:
```diff
@@ -821,6 +821,76 @@ exports[`KeyboardShortcutHelpModal should match snapshot when the modal is activ
+              <kbd
+                class="edit-post-keyboard-shortcut-help-modal__shortcut-key"
+              >
+                Shift
```
Ctrl?

scruffian commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/test/__snapshots__/index.js.snap:
```diff
@@ -821,6 +821,76 @@ exports[`KeyboardShortcutHelpModal should match snapshot when the modal is activ
+            class="edit-post-keyboard-shortcut-help-modal__shortcut-term"
+          >
+            <kbd
+              aria-label="Shift + Alt + 1 6"
```
Ctrl?

scruffian commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/test/__snapshots__/index.js.snap:
```diff
@@ -821,6 +821,76 @@ exports[`KeyboardShortcutHelpModal should match snapshot when the modal is activ
+              <kbd
+                class="edit-post-keyboard-shortcut-help-modal__shortcut-key"
+              >
+                Shift
```
Ctrl?

draganescu replied:
`access: ( _isApple ) => ( _isApple() ? [ CTRL, ALT ] : [ SHIFT, ALT ] ),`

The tests run on a linux env so the snapshots match :)

scruffian reviewed (approved):
LGTM.

✅ tested converting paragraphs to heading 1, 2, 3, 4, 5, 6
✅ tested converting headings to a paragraph

t-hamano commented:
@draganescu 
What do you think about adding this feature to other editors (edit-site/edit-widget)? If it seems reasonable, I'd like to work on it, including E2E testing.

draganescu commented:
@t-hamano  sure thing, I think it's a cool addition from a consistency point of view. The intended goal here is to improve the writing experience via easier formatting - in the widget and site editors we expect more of a structural approach and less focus on flow, but sure, even cycling through heading sizes with a KBD shortcut is awesome.

Add me to review! 🙇🏻
